src/parpeoptimization/optimizationProblem.cpp

```diff
@@ -54,6 +54,84 @@ void *getLocalOptimumThreadWrapper(void *optimizationProblemVp) {
     return result;
 }
 
+void optimizationProblemGradientCheckMultiEps(OptimizationProblem *problem,
+                                      int numParameterIndicesToCheck
+                                      ) {//perhaps list of epsilons?
+    // set eps (if no default added yet)
+    std::vector<double> multi_eps {1e-1, 1e-3, 1e-5, 1e-7, 1e-9};
+
+    // setting the number of parameters to the minimum of
+    // numParamaterIndicesToCheck and dimension of the problem
+    int numParameters = problem->cost_fun_->numParameters();
+    numParameterIndicesToCheck =
+        std::min(numParameterIndicesToCheck, numParameters);
+    // choose random parameters to check
+    std::vector<int> parameterIndices(numParameters); // ask Daniel
+    std::iota(parameterIndices.begin(), parameterIndices.end(), 0);
+    std::random_device rd;
+    std::mt19937 g(rd());
+
+    optimizationProblemGradientCheckMultiEps(problem, parameterIndices, multi_eps);
+}
+
+void optimizationProblemGradientCheckMultiEps(OptimizationProblem *problem,
+                                              gsl::span<const int> parameterIndices,
+                                              gsl::span<double> multi_eps){
+    //get a random theta
+    std::vector<double> theta(problem->cost_fun_->numParameters());
+    problem->fillInitialParameters(theta);
+    double fc = 0; // f(theta)
+    //evaluate the objective function at theta and get analytical gradient
+    std::vector<double> gradient(theta.size());
+    problem->cost_fun_->evaluate(theta, fc, gradient);
+
+    std::vector<double> thetaTmp(theta);
+    // loop through indices, then loop trough epsilons
+    for(int curInd : parameterIndices) {
+        double eps_best=0, regRelErr_best=0, absErr_best=0, fd_c_best=0;
+
+        // getting the analytical gradient of current index
+        double curGrad = gradient[curInd];
+
+        for(double epsilon : multi_eps) {
+            double fb = 0, ff = 0; // f(theta + eps) , f(theta - eps)
+
+            thetaTmp[curInd] = theta[curInd] + epsilon;
+            problem->cost_fun_->evaluate(gsl::span<double>(thetaTmp), ff,
+                                         gsl::span<double>());
+
+            thetaTmp[curInd] = theta[curInd] - epsilon;
+            problem->cost_fun_->evaluate(gsl::span<double>(thetaTmp), fb,
+                                         gsl::span<double>());
+            // calculating the finite difference
+            double fd_c = (ff - fb) / (2 * epsilon);
+            //reverting thetaTmp back to original
+            thetaTmp[curInd] = theta[curInd];
+
+            double reg = 1e-5;
```

Comment:
Shall we move `epsilon` here? Didn't really test how big the influence is. (That value was chosen rather arbitrarily).

Reply:
in python epsilon was used, so would say yes.